pkg/private/archive.py

```diff
@@ -252,25 +315,17 @@ def add_file(self,
       return
     if name == '.':
       return
-    name = self.add_root(name)
-    # Do not add a directory that is already in the tar file.
+    name = self.add_root_prefix(name)
     if kind == tarfile.DIRTYPE and name in self.directories:
       return
+    if name in self.members:
```

Comment:
Do we need to consider a directory and a file of same name may co-exist? If not, we can delete L319~L320, where we check duplicate directories because we'll check if it exist in self.members. If so, we'll need to check `kind` in this if.

Reply:
A directory and file of the same name can never coexist. 

Those 2 lines can be deleted. FWIW, I wanted to eliminate self.directories, but there are tests which peer into this class and look at self.members. In order to keep a promise that this PR only changed the implementation, and not the external behavior, I did not do that here.  I may in a future cleanup.